pkg/resourceinterpreter/default/native/retain.go

```diff
@@ -40,34 +41,65 @@ func retainPodFields(desired, observed *unstructured.Unstructured) (*unstructure
+	// `spec.terminationGracePeriodSeconds` (allow it to be set to 1 if it was previously negative).
+	// More info: https://kubernetes.io/docs/concepts/workloads/pods/#pod-update-and-replacement
+	//
+	// There are many fields in Pod, but only a few fields that can be modified. We choose to return clusterPod
+	// because it is unnecessary waste to copy from clusterPod to desiredPod.
+	clusterPod.Labels = desiredPod.Labels
+	clusterPod.Annotations = desiredPod.Annotations
+	updatePodImages(desiredPod, clusterPod)
+	updatePodActiveDeadlineSeconds(desiredPod, clusterPod)
+	updatePodTolerations(desiredPod, clusterPod)
+
+	unstructuredObj, err := helper.ToUnstructured(clusterPod)
+	if err != nil {
+		return nil, fmt.Errorf("failed to transform Pod: %v", err)
+	}
+
+	return unstructuredObj, nil
+}
+
+func updatePodImages(desiredPod, clusterPod *corev1.Pod) {
+	for clusterIndex, clusterContainer := range clusterPod.Spec.InitContainers {
+		for _, desiredContainer := range desiredPod.Spec.InitContainers {
 			if desiredContainer.Name == clusterContainer.Name {
-				desiredPod.Spec.Containers[desiredIndex].VolumeMounts = clusterContainer.VolumeMounts
+				clusterPod.Spec.InitContainers[clusterIndex].Image = desiredContainer.Image
 				break
 			}
 		}
 	}
-
-	// retain volumeMounts in each init container
-	for _, clusterInitContainer := range clusterPod.Spec.InitContainers {
-		for desiredIndex, desiredInitContainer := range desiredPod.Spec.InitContainers {
-			if desiredInitContainer.Name == clusterInitContainer.Name {
-				desiredPod.Spec.InitContainers[desiredIndex].VolumeMounts = clusterInitContainer.VolumeMounts
+	for clusterIndex, clusterContainer := range clusterPod.Spec.Containers {
+		for _, desiredContainer := range desiredPod.Spec.Containers {
+			if desiredContainer.Name == clusterContainer.Name {
+				clusterPod.Spec.Containers[clusterIndex].Image = desiredContainer.Image
 				break
 			}
 		}
 	}
-	unstructuredObj, err := helper.ToUnstructured(desiredPod)
-	if err != nil {
-		return nil, fmt.Errorf("failed to transform Pod: %v", err)
+}
+
+func updatePodActiveDeadlineSeconds(desiredPod, clusterPod *corev1.Pod) {
+	desiredActiveDeadlineSeconds := pointer.Int64Deref(desiredPod.Spec.ActiveDeadlineSeconds, 0)
+	clusterActiveDeadlineSeconds := pointer.Int64Deref(clusterPod.Spec.ActiveDeadlineSeconds, 0)
+	if (clusterPod.Spec.ActiveDeadlineSeconds == nil && desiredActiveDeadlineSeconds > 0) ||
+		(desiredActiveDeadlineSeconds > 0 && desiredActiveDeadlineSeconds < clusterActiveDeadlineSeconds) {
+		clusterPod.Spec.ActiveDeadlineSeconds = desiredPod.Spec.ActiveDeadlineSeconds
 	}
+}
 
-	return unstructuredObj, nil
+func updatePodTolerations(desiredPod, clusterPod *corev1.Pod) {
+outer:
+	for _, desiredToleration := range desiredPod.Spec.Tolerations {
+		for _, clusterToleration := range clusterPod.Spec.Tolerations {
+			if clusterToleration == desiredToleration {
+				continue outer
```

Comment:
Can we achieve the same goal by adding a variable, like `xxFlag`?

Reply:
Could you explain to me why we need to do this, or what is the benefit compared to label loop?

Reply:
Personally, I haven't used the label loop before, and it feels like the code is hard to read.
Ask @RainbowMango to help take a look.
/cc @RainbowMango

Reply:
OK. I will check it as soon as possible.